src/OrbitBase/File.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitBase/File.h"
+
+#include "OrbitBase/Logging.h"
+
+namespace orbit_base {
+
+static constexpr auto FdDeleter = [](int fd) {
+  if (fd != -1) close(fd);
+};
+
+ErrorMessageOr<unique_fd> OpenFileForReading(const std::filesystem::path& path) {
+#if defined(__linux)
+  constexpr int open_flags = O_RDONLY | O_CLOEXEC;
+#elif defined(_WIN32)
+  constexpr int open_flags = O_RDONLY | O_BINARY;
+#endif  // defined(__linux)
+  int fd = TEMP_FAILURE_RETRY(open(path.string().c_str(), open_flags));
+  if (fd == -1) {
+    return ErrorMessage(
+        absl::StrFormat("Unable to open file \"%s\": %s", path.string(), SafeStrerror(errno)));
+  }
+
+  return unique_fd{fd, FdDeleter};
+}
+
+ErrorMessageOr<unique_fd> OpenFileForWriting(const std::filesystem::path& path) {
+#if defined(__linux)
+  constexpr int open_flags = O_WRONLY | O_CREAT | O_TRUNC | O_CLOEXEC;
+  constexpr int open_mode = 0600;
+#elif defined(_WIN32)
+  constexpr int open_flags = O_WRONLY | O_CREAT | O_TRUNC | O_BINARY;
+  constexpr int open_mode = _S_IREAD | _S_IWRITE;
+#endif  // defined(__linux)
+  int fd = TEMP_FAILURE_RETRY(open(path.string().c_str(), open_flags, open_mode));
+
+  if (fd == -1) {
+    return ErrorMessage(
+        absl::StrFormat("Unable to open file \"%s\": %s", path.string(), SafeStrerror(errno)));
+  }
+
+  return unique_fd{fd, FdDeleter};
+}
+
+ErrorMessageOr<void> WriteFully(const unique_fd& fd, std::string_view content) {
```

Comment:
Don't you want to also support passing a bare file descriptor (`int`) instead of exclusively a `unique_fd`?

Reply:
yes - this one is on purpose, we should avoid using bare fds in our code.

Reply:
Hhm, I don't have a strong opinion nor I see it being a problem, but conceptually I think this is the wrong place to enforce it.
At least from the name, `unique_fd` seems to serve the purpose of resource management of file descriptors, not the purpose of completely hiding the presence of file descriptors (which is something that `FILE` would do). 
Compare with `unique_ptr`: you want to use them, but it makes no sense for a function to accept `const unique_ptr&`, you could just pass the pointer. This is the same: you want to *create* `unique_fd`s, but it still makes sense to pass the bare file descriptor around.

Reply:
well, unique_ptr is not something we want to enforce because we do not want to decide for the user how they do memory management, is it unique_ptr, shared_ptr, weak_ptr or something else. For this we kind of want to not have bare fds in the code for files. I would default on unique_fd and if an reasonable use case requiring bare fd comes up we can change it.

Reply:
Yes, we can always change it. One thing I was thinking is that in theory we could just make the fact that `unique_fd` is managing a file descriptor completely opaque. For example, it could just be `File` or `FileHandle`. It would fit better if this is meant to become a completely independent library to read and write files. I don't know if it's necessarily better though.